solutions/go/negative_character_groups/code/cmd/mygrep/main.go

```diff
@@ -0,0 +1,72 @@
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+	"unicode"
+	"unicode/utf8"
+)
+
+// Usage: echo <input_text> | your_grep.sh -E <pattern>
+func main() {
+	if len(os.Args) < 3 || os.Args[1] != "-E" {
+		fmt.Fprintf(os.Stderr, "usage: mygrep -E <pattern>\n")
+		os.Exit(2) // 1 means no lines were selected, >1 means error
+	}
+
+	pattern := os.Args[2]
+
+	line, err := io.ReadAll(os.Stdin) // assume we're only dealing with a single line
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "error: read input text: %v\n", err)
+		os.Exit(2)
+	}
+
+	ok, err := matchLine(line, pattern)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "error: %v\n", err)
+		os.Exit(2)
+	}
+
+	if !ok {
+		os.Exit(1)
+	}
+
+	// default exit code is 0 which means success
+}
+
+func matchLine(line []byte, pattern string) (bool, error) {
+	switch {
+	case pattern == `\d`:
+		for _, char := range string(line) { // ranging over string handles utf-8 multibyte characters correctly
+			if unicode.IsDigit(char) { // char is rune not byte
+				return true, nil
+			}
+		}
+
+		return false, nil
+	case pattern == `\w`:
+		for _, char := range string(line) {
+			if unicode.IsLetter(char) {
+				return true, nil
+			}
+		}
+
+		return false, nil
+	case strings.HasPrefix(pattern, "[^") && strings.IndexByte(pattern, ']') == len(pattern)-1: // starts with "[^" and the ']' is the last
```

Comment:
Can we use `strings.HasSuffix` so that it's symmetric with `strings.HasPrefix`?

Reply:
Nope, we are checking not the end of pattern, but the end of group, which ends on the first ']' character. So the difference will be on a pattern `[abc]$%]`, for example.

Reply:
I added the comment, but looks like it's not explanatory enough. Maybe change it to ?
`// starts with "[^" and the ']' is the last. Note the [abc]%-] example.`

Reply:
Ah, so we're looking to see if the first occurrence of `]` is the last character. Sorry, I was just quickly glancing through. 

I see the case you're trying to handle here, I'm not sure that we need to support it though. The implementation we'll use here will obviously be incomplete since we haven't hit "combining_character_classes" yet. `[abc]de` isn't a valid pattern at this stage. At this stage, it's expected that only a single character class will be provided (i.e. if the pattern starts with `[`, it's going to contain characters and end with `]`). 

That said, we will need to handle this in the future for sure, so maybe we should use whatever implementation can be re-used effectively in the `combining_character_classes` stage?

Another reason I want to avoid this: the "checking for last occurrence of ]" isn't necessarily correct, you could have `[a\]c]` as a pattern, for example. 

<img width="161" alt="Screenshot 2022-10-28 at 11 56 42" src="https://user-images.githubusercontent.com/3893573/198571427-caeb4133-80a5-429d-aefe-03bb93a3e6ee.png">

(not saying we have to handle this! Just pointing out a valid failure case)

Reply:
So I can just check we have ']' in the end with `strings.HasSuffix`. And we are fine it will only work on correct inputs.

I can do the same as in the python solution, we are fine we accept patterns, which are invalid.

I can add a simple implementation of finding closing bracket with skipping escaped symbols with a loop. We can reuse it later and try to cover all common patterns to work correctly. We won't be able to cover actually all the same behaviour, I think, that would be too complex.

Or something in between.

Reusing this code doesn't look very valuable here, because it changes dramatically anyway on combining classes stage. And some input is incorrect on character groups stage (and we should report that), but is correct on combining classes stage.

Which approach is preferable? I would go with the more complex and closer to the actual regexp since it's not the first stage, there are couple of very simple stages before. So here we can add one a bit more complex. Implement the login in a separate function and reuse it later.

Reply:
I actually hadn't thought of escaped symbols at all when working on the tester. Agree that this is something we should probably handle... Will think a bit and respond.

Reply:
any updates?

Reply:
Hey - yep I'm going to update the tester to include escaped symbols. Once that's done, I'll get the Python solution updated and then we can revisit this based on that.

Let's hold off on the work here for a couple of days (can work on the GitHub action in the meantime) until all the above mentioned pieces are in place

Reply:
@nikandfor I've updated the Python solutions in https://github.com/codecrafters-io/build-your-own-grep/pull/13. 

I initially intended to add support for escaped symbols, but that didn't go well - the only "escaped" characters we could use in the "positive_character_groups" stage was something like `\d`, and that in turn would require the learner to implement recursion (essentially getting to do work very similar to the `combining_character_classes` stage). 

We of course want to cover all kinds of patterns, but this level of complexity seems like a better candidate for [challenge extensions](https://app.codecrafters.io/vote/challenge-extension-ideas?course=redis) than the core challenges. 

So I kept the existing tests (along with the ones you suggested to add), and just changed the Python solution to handle them.

As for the code changes here, I think the simplest solution would be something like this: 

```go
case strings.HasPrefix(pattern, "["):
		charset := pattern[1 : strings.IndexByte(pattern, "]")-1]
		return bytes.ContainsAny(line, charset), nil
```

This won't work with escaped `]` characters inside the group, but that's okay. It also won't complain on invalid patterns - that's fine too. We can spin out "error messages" into a separate challenge extension later.

Reply:
@nikandfor also, let's get to updating this _after_ the GitHub actions work is done, so that we can test it out on this PR 🙂